### PR TITLE
Fixes #3097. Now that View is IDisposable, it should expose a Disposing event.

### LINKED ISSUE #3097: Now that View is IDisposable, it should expose a Disposing event
In v2, View is IDisposable.

However, there's no way, at present, for anything dependent on a view to know when that view has been disposed or is being disposed, other than trying something that's illegal at runtime, situationally.

This can lead to memory leaks, runtime exceptions, and other unexpected behavior, in consumers.

Proposal:

Implement a standard Disposing event, to be invoked in the Dispose method, so that anything dependent on a given View can gracefully clean up, when that view is disposed.

One place this is already relevant is in TerminalGuiDesigner: https://github.com/gui-cs/TerminalGuiDesigner/issues/270

Having such an event _does_ expose the possibility of bad consumer code blocking during Dispose, but that's _their_ fault, not Terminal.Gui's. We could attempt to protect users from themselves by implementing some sort of deadline mechanism, but I don't think that's necessary and has more potential downsides than benefits (plus the complexity of implementing it in the first place).

### PR DESCRIPTION
**Pull request by BDisp**
Fixes #3097 - Added `Disposing` event which will help to take the necessary measure when needed.

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

File: Terminal.Gui/Input/Responder.cs [hunk 1]
```diff
@@ -47,6 +47,11 @@ public Responder ()
 	}
 #endif
 
+	/// <summary>
+	/// Invokes subscribers that this is disposed.
```
Comment (tig) [inline Terminal.Gui/Input/Responder.cs:51]: `/// Event raised when <see cref="Dispose"/> is called to signal that this object has been disposed.`

File: Terminal.Gui/Input/Responder.cs [hunk 2]
```diff
@@ -187,6 +192,7 @@ public void Dispose ()
 	{
 		// Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
 		Dispose (disposing: true);
+		Disposing?.Invoke (this, EventArgs.Empty);
```
Comment (tig) [inline Terminal.Gui/Input/Responder.cs:195]: I think this line should be moved to after the `wasDisposed = true` line.

File: UnitTests/Input/ResponderTests.cs [hunk 3]
```diff
@@ -112,11 +113,11 @@ public void IsOverridden_True_IfOverridden ()
 		// MouseEvent is defined on Responder IS overriden on ScrollBarView (but not View)
 		Assert.True (Responder.IsOverridden (new ScrollBarView () { Text = "ScrollBarView overrides MouseEvent" }, "MouseEvent"));
 
-		//// OnKeyDown is defined on View
-		//Assert.True (Responder.IsOverridden (new View () { Text = "View overrides OnKeyDown" }, "OnKeyDown"));
+		// OnKeyDown is defined on View
```
Comment (tig) [inline UnitTests/Input/ResponderTests.cs:116]: FWIW, this IsOverriden hack must be removed in v2. It should be removed as part of #https://github.com/gui-cs/Terminal.Gui/issues/3029

File: UnitTests/Input/ResponderTests.cs [hunk 4]
```diff
@@ -129,4 +130,124 @@ public void IsOverridden_True_IfOverridden ()
...
 #endif
 	}
+
+	[Fact]
+	public void Responder_Not_Notifying_Dispose ()
+	{
+		List<View> views = new List<View> ();
+		var container1 = new View () { Id = "Container1" };
+
+		for (int i = 0; i < 20; i++) {
```
Comment (tig) [inline UnitTests/Input/ResponderTests.cs:140]: Is there really any value in doing these tests 20 times vs just once? 

This seems like un-needed complexity in the unit tests.

If there is a specific edge case that needs to be tested, then a specific unit test that tests that should be provided.

File: UnitTests/Input/ResponderTests.cs [hunk 5]
```diff
@@ -129,4 +130,124 @@ public void IsOverridden_True_IfOverridden ()
 		Assert.Empty (Responder.Instances);
 #endif
 	}
+
```
Comment (tig) [inline UnitTests/Input/ResponderTests.cs:133]: Is there value in having a version of these unit tests that tests all view sub-classes?

File: Terminal.Gui/Input/Responder.cs (hunk 2 above)
Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs:195]: Ah I only saw this now, in this case `Disposed` is the right name.

File: UnitTests/Input/ResponderTests.cs (hunk 3 above)
Comment (BDisp) [reply UnitTests/Input/ResponderTests.cs:116]: Ok I leave as is and then per #3029 will be removed, right?

File: UnitTests/Input/ResponderTests.cs (hunk 5 above)
Comment (BDisp) [reply UnitTests/Input/ResponderTests.cs:133]: I think so, because I'm testing 3 cases:
1. Without disposed notification.
2. With disposed notification only on the second container.
3. With disposed notification only on the first container.

File: UnitTests/Input/ResponderTests.cs (hunk 4 above)
Comment (BDisp) [reply UnitTests/Input/ResponderTests.cs]: I'll change it to only test a single view for all.

File: Terminal.Gui/Input/Responder.cs (hunk 2 above)
Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs]: Done.

File: Terminal.Gui/Input/Responder.cs [hunk 6]
```diff
@@ -47,6 +47,11 @@ public Responder ()
 	}
 #endif
 
+	/// <summary>
+	/// Invokes subscribers that this is disposed.
+	/// </summary>
+	public event EventHandler Disposing;
```
[the remark this answers is not in the document]
Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs]: Done.

File: Terminal.Gui/Input/Responder.cs (hunk 2 above)
Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs]: @tig we need to resolve this. I need to invoke the event before the `GC.SuppressFinalize` call. Then I need your finally opinion of the event name, `Disposing` or `Disposed`?

File: UnitTests/Input/ResponderTests.cs (hunk 5 above)
Comment (BDisp) [reply UnitTests/Input/ResponderTests.cs:138]: Ak I think your talking about why not only using the base class `Responder`. The reason was I wanted to use `View` properties, like the `Id`, `Subviews`, etc...

File: Terminal.Gui/Input/Responder.cs (hunk 2 above)
Comment (tig) [reply Terminal.Gui/Input/Responder.cs]: > @tig we need to resolve this. I need to invoke the event before the `GC.SuppressFinalize` call. Then I need your finally opinion of the event name, `Disposing` or `Disposed`?

Explain why please.

Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs]: Although the event is invoked on the `Dispose` method it must be called before calling the `GC.SuppressFinalize` to avoid unit test failure. So it isn't already disposed, although the `View` already disposed others objects. So do you confirm that you want the event named as `Disposed`. If affirmative then it's already changed.

Comment (tig) [reply Terminal.Gui/Input/Responder.cs]: > Although the event is invoked on the `Dispose` method it must be called before calling the `GC.SuppressFinalize` to avoid unit test failure. So it isn't already disposed, although the `View` already disposed others objects. So do you confirm that you want the event named as `Disposed`. If affirmative then it's already changed.

Are you sure the unit test is valid?

Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs]: Of course it's. Now it only handle one view as you asked. But if you think it isn't valid please say why you are think that.

Comment (tig) [reply Terminal.Gui/Input/Responder.cs]: > Of course it's. Now it only handle one view as you asked. But if you think it isn't valid please say why you are think that.

Sorry, but I don't have time to dive deeply into this and understand it myself. I'd like you to explain to me what fails if the event is fired after `GC.SuppresFinalize`. With my limited understanding it should make any difference. 

I can't answer on what the event should be named without understanding this.

Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs]: Because when the event is invoked `container1` was already had disposed the view and the `container2` could not dispose all the `Responder.Instances`. See the error below. But this failure is arbitrary. Before I push the commit all the unit tests passed locally but failed on the server after pushed it.

```
  Failed Terminal.Gui.InputTests.ResponderTests.Disposed_Event_Notify_All_Subscribers_On_The_First_Container [2 ms]
  Error Message:
   Assert.Empty() Failure: Collection was not empty
Collection: [FrameView()(0,0,0,0), Frame(Margin)(0,0,0,0), Frame(Border)(0,0,0,0), Frame(Padding)(1,1,0,0), FrameView(Test)(0,0,0,0), ···]
  Stack Trace:
     at Terminal.Gui.InputTests.ResponderTests.Disposed_Event_Notify_All_Subscribers_On_The_First_Container() in D:\a\Terminal.Gui\Terminal.Gui\UnitTests\Input\ResponderTests.cs:line 238
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
  Failed Terminal.Gui.InputTests.ResponderTests.Responder_Not_Notifying_Dispose [< 1 ms]
  Error Message:
   Assert.Empty() Failure: Collection was not empty
Collection: [FrameView()(0,0,0,0), Frame(Margin)(0,0,0,0), Frame(Border)(0,0,0,0), Frame(Padding)(1,1,0,0), FrameView(Test)(0,0,0,0), ···]
  Stack Trace:
     at Terminal.Gui.InputTests.ResponderTests.Responder_Not_Notifying_Dispose() in D:\a\Terminal.Gui\Terminal.Gui\UnitTests\Input\ResponderTests.cs:line 168
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

Comment (tig) [reply Terminal.Gui/Input/Responder.cs]: I still don't understand. 

However, I now think that the event should be `Disposing` and should be fired before `protected virtual void Dispose (bool disposing)` is called.

Comment (tig) [reply Terminal.Gui/Input/Responder.cs]: The API doc should read:

`/// Event raised when <see cref="Dispose()"/> has been called to signal that this object is being disposed.`

Comment (BDisp) [reply Terminal.Gui/Input/Responder.cs:194]: > I still don't understand.

Me either 😃 but I can't debug it because I don't get any fail locally. Some racing condition caused by the event being called after `GC.SuppressFinalize` which depend of the OS to do the cleanup from the garbage collector.

**Review by tig — APPROVED** [review]
